mmrazor/models/algorithms/distill/configurable/self_distill.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import List, Optional
+
+import torch
+from mmengine import BaseDataElement
+from torch import nn
+
+from mmrazor.models.utils import add_prefix
+from mmrazor.registry import MODELS
+from ...base import BaseAlgorithm, LossResults
+
+
+@MODELS.register_module()
+class SelfDistill(BaseAlgorithm):
+    """``SingleTeacherDistill`` can be used to develop distill algorithms
+    without teacher.
+
+    Args:
+        distiller (dict): The config dict for built distiller. Distiller may
+            have teacher.
+        student_trainable (bool): Whether the student is trainable. Defaults
+            to True.
+        calculate_student_loss (bool): Whether to calculate student loss
+            (original task loss) to update student model. Defaults to True.
+    """
+
+    def __init__(self,
+                 distiller: dict,
+                 student_trainable: bool = True,
+                 calculate_student_loss: bool = True,
+                 **kwargs) -> None:
+        super().__init__(**kwargs)
+
+        self.distiller = MODELS.build(distiller)
+        # The student model will not calculate gradients and update parameters
+        # in some pretraining process.
+        self.student_trainable = student_trainable
+
+        # The student loss will not be updated into ``losses`` in some
+        # pretraining process.
+        self.calculate_student_loss = calculate_student_loss
+
+        # In ``ConfigurableDistller``, the recorder manager is just
+        # constructed, but not really initialized yet.
+        self.distiller.prepare_from_student(self.student)
+        # Still prepare from self-teacher. Teacher recorders of
+        # ``SelfDistiller`` hook from self.student but require detach().
+        self.distiller.prepare_from_teacher(self.student)
```

Comment:
It seems that line 45 and line 48 are the same.

Reply:
Line 45 prepare `student_recorders` on student model with `prepare_from_student()`
Line 48 prepare `teacher_recorders` on student model with `prepare_from_teacher()`
In `SelfDistiller`, `teacher_recorders` record self-teacher vars on student model and require detach() in calculation.

Reply:
My bad!